### PR TITLE
prov/gni: Implemented fi direct.

### LINKED ISSUE #600: Implement fi_direct

### PR DESCRIPTION
**Pull request by e-harvey**
- Added missing gnix functions.
- Added DIRECT_FN and STATIC to the beginning of GNI
  API functions to make those functions globally visible
  when compiling with fi direct enabled.
- Updated Makefile.include to include gnix_poll.*.
- Added symbol version info for the gni provider.

Signed-off-by: Evan Harvey eharvey@lanl.gov

@hppritcha @sungeunchoi @jswaro @jshimek @tonyzinger @chuckfossen 

Fixes #600

### CODE REVIEW

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by e-harvey** [comment]
Perhaps we could add '**attribute**(always_inline)' to the 'DIRECT_FN' macro defined in fi_prov.h.  Will this force inlining when compiling with clang?

**Comment by hppritcha** [comment]
not sure, but probably its worth adding this attribute to the function declarations.

**Comment by sungeunchoi** [comment]
Yes newer version support always_inline (don't know how new it has to be).  That said, we're not building with clang on our machines and currently don't have any plans to do so.

**Comment by hppritcha** [comment]
@e-harvey i renamed the include file from the provider that goes in to libfabric.map in response to sean's concerns about ofiwg/libfabric#1740

**Comment by e-harvey** [comment]
@hppritcha Ok, I renamed provider.map -> provider_FABRIC_1.0.map.

On Wed, 24 Feb 2016 10:04:59 -0800
Howard Pritchard notifications@github.com wrote:

> @e-harvey i renamed the include file from the provider that goes in to libfabric.map in response to sean's concerns about ofiwg/libfabric#1740
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/ofi-cray/libfabric-cray/pull/671#issuecomment-188385232

## 

Evan Harvey

**Comment by e-harvey** [comment]
Rebased with maser, moved DIRECT_FN and STATIC macros in gnix.h, and added gnix_fabric_trywait.  Please review.

File: prov/gni/src/gnix_ep.c [hunk 1]
```diff
@@ -634,20 +655,43 @@ static ssize_t gnix_ep_tsendmsg(struct fid_ep *ep,
...
+ * @param[in] buf	the data to send
+ * @param[in] len	the length of buf
+ * @param[in] data	remote CQ data to transfer with the data from buf
+ * @param[in] dest_addr	the desitnation address for connectionless transfers
+ * @param[in] tag	the tag associated with the message
+ *
+ * @return FI_SUCCESS	upon successfully writing to the destination
+ * @return -FI_ERRNO	upon an error
+ * @return -FI_ENOSYS	if this operation is not supported
+ */
```
Comment (hppritcha) [inline prov/gni/src/gnix_ep.c:687]: thanks for adding the doxygen style comments

**Comment by hppritcha** [comment]
:+1:

**Pull request merged**